File: ejemplos/nerve/08-puente-http/ejemplo.py
"""
Ejemplo: Servidor HTTP con FastAPI

Qué enseña este ejemplo:
  - Crear una aplicación básica de FastAPI.
  - Definir un modelo de datos (Pydantic) para recibir JSON.
  - Crear un endpoint (ruta) POST que lea esos datos.

Para tu reto (Puente HTTP para Nerve):
  Cuando el endpoint reciba los datos, deberás usar el cliente Nerve (app.state.nerve)
  para retransmitir la información hacia el hub de Nerve.
"""

from fastapi import FastAPI
from pydantic import BaseModel

# 1. Definir el modelo de los datos que esperamos recibir
# Esto asegura que quien envíe el POST, nos mande un JSON con un campo "mensaje"
class MensajePayload(BaseModel):
    mensaje: str

# 2. Inicializar la aplicación FastAPI
app = FastAPI()

# 3. Ruta GET básica para comprobar que funciona
@app.get("/")
def leer_raiz():
    return {"status": "Servidor funcionando"}

# 4. Ruta POST que será nuestro "puente"
# FastAPI automáticamente convierte el cuerpo del POST desde JSON hacia nuestro modelo
@app.post("/enviar")
def recibir_y_reenviar(payload: MensajePayload):
    # En la terminal veremos lo que llegó
    print(f"📩 Recibido por HTTP: {payload.mensaje}")
    
    # ¡Aquí es donde integrarías Nerve! 
    # Ejemplo: cliente_nerve.publicar("canal_http", payload.mensaje)
    
    return {"status": "ok", "recibido": payload.mensaje}

# NOTA: Para ejecutar esto se usa uvicorn desde la terminal:
# uvicorn ejemplo:app --reload

File: ejemplos/nerve/08-puente-http/ejemplo.rs
/**
 * Ejemplo: Servidor HTTP con Axum
 * 
 * Qué enseña este ejemplo:
 *   - Crear un servidor HTTP moderno y rápido en Rust usando axum.
 *   - Usar 'serde' para deserializar un cuerpo JSON automáticamente.
 *   - Configurar una ruta POST y enviar una respuesta JSON.
 * 
 * Para tu reto (Puente HTTP para Nerve):
 *   Cuando la ruta reciba los datos, deberás usar el cliente Nerve 
 *   para retransmitir la información hacia el hub de Nerve. Para compartir
 *   el cliente Nerve con Axum, necesitarás usar el sistema de "State" de Axum
 *   con un Arc<Mutex<...>>.
 */

use axum::{
    routing::{get, post},
    Router,
    Json,
};
use serde::{Deserialize, Serialize};

// 1. Estructura para lo que ESPERAMOS recibir
// 'Deserialize' permite a Axum convertir automáticamente el JSON entrante a este struct
#[derive(Deserialize)]
struct MensajePayload {
    mensaje: String,
}

// 2. Estructura para lo que VAMOS a responder
// 'Serialize' permite a Axum convertir el struct a JSON para responder
#[derive(Serialize)]
struct RespuestaPayload {
    status: String,
    recibido: String,
}

#[tokio::main]
async fn main() {
    // 3. Crear el enrutador de la aplicación y definir las rutas
    let app = Router::new()
        .route("/", get(ruta_raiz))
        .route("/enviar", post(recibir_y_reenviar));

    // 4. Iniciar el servidor
    let puerto = "0.0.0.0:3000";
    let listener = tokio::net::TcpListener::bind(puerto).await.unwrap();
    
    println!("Servidor Rust (Axum) escuchando en http://localhost:3000");
    axum::serve(listener, app).await.unwrap();
}

// Ruta GET básica
async fn ruta_raiz() -> &'static str {
    "Servidor funcionando"
}

// 5. Ruta POST que será nuestro "puente"
// Axum se encarga de verificar que el POST tenga un JSON válido y lo mapea a `payload`
async fn recibir_y_reenviar(Json(payload): Json<MensajePayload>) -> Json<RespuestaPayload> {
    
    // En la terminal veremos lo que llegó
    println!("📩 Recibido por HTTP: {}", payload.mensaje);

    // ¡Aquí es donde integrarías Nerve! 
    // Ejemplo: state.cliente_nerve.publicar("canal_http", &payload.mensaje).await;

    // Retornamos un JSON de respuesta
    let respuesta = RespuestaPayload {
        status: "ok".to_string(),
        recibido: payload.mensaje,
    };

    Json(respuesta)
}
